tests/phpunit/PaymentMethods/WC_Stripe_Express_Checkout_Helper_Test.php

```diff
@@ -984,4 +985,79 @@ public function provide_test_has_free_trial() {
 			],
 		];
 	}
+
+	/**
+	 * Test that OPC detection logic works correctly.
+	 *
+	 * @dataProvider provide_opc_detection_scenarios
+	 *
+	 * @param bool  $is_opc Whether the page is detected as One Page Checkout.
+	 * @param array $button_locations Button location settings.
+	 * @param bool  $expected Expected result for should_show_express_checkout_button.
+	 * @group opc
+	 *
+	 * @return void
+	 */
+	public function test_opc_detection_logic( $is_opc, $button_locations, $expected ) {
+		$stripe_settings                                     = WC_Stripe_Helper::get_stripe_settings();
+		$stripe_settings['payment_request_button_locations'] = $button_locations;
+		WC_Stripe_Helper::update_main_stripe_settings( $stripe_settings );
+
+		$gateway = $this->getMockBuilder( WC_Stripe_UPE_Payment_Gateway::class )
+			->disableOriginalConstructor()
+			->getMock();
+
+		$wc_stripe_ece_helper_mock = $this->createPartialMock(
+			WC_Stripe_Express_Checkout_Helper::class,
+			[
+				'is_one_page_checkout',
+				'is_product',
+				'is_checkout',
+				'allowed_items_in_cart',
+				'get_product',
+			],
+			[ $gateway ]
+		);
+
+		// Create a mock product.
+		$product = WC_Helper_Product::create_simple_product();
+		$is_product_page = $is_opc || in_array( 'product', $button_locations, true );
+
+		// Mock the methods.
+		$wc_stripe_ece_helper_mock->expects( $this->any() )->method( 'is_one_page_checkout' )->willReturn( $is_opc );
+		$wc_stripe_ece_helper_mock->expects( $this->any() )->method( 'is_product' )->willReturn( $is_product_page );
+		$wc_stripe_ece_helper_mock->expects( $this->any() )->method( 'is_checkout' )->willReturn( false );
+		$wc_stripe_ece_helper_mock->expects( $this->any() )->method( 'allowed_items_in_cart' )->willReturn( true );
+		$wc_stripe_ece_helper_mock->expects( $this->any() )->method( 'get_product' )->willReturn( $is_product_page ? $product : false );
```

Comment:
As above, can we remove the `expects()` calls that are effectively noise?
```suggestion
		$wc_stripe_ece_helper_mock->method( 'is_one_page_checkout' )->willReturn( $is_opc );
		$wc_stripe_ece_helper_mock->method( 'is_product' )->willReturn( $is_product_page );
		$wc_stripe_ece_helper_mock->method( 'is_checkout' )->willReturn( false );
		$wc_stripe_ece_helper_mock->method( 'allowed_items_in_cart' )->willReturn( true );
		$wc_stripe_ece_helper_mock->method( 'get_product' )->willReturn( $is_product_page ? $product : false );
```

Reply:
Thanks! Removed in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4757/commits/2270db97baf30b4119f1edcedc2d3e17c12a3a93